pykokkos/core/runtime.py

```diff
@@ -129,13 +132,69 @@ def run_workunit(
                 raise RuntimeError("ERROR: operation cannot be None for Debug")
             return run_workunit_debug(policy, workunit, operation, initial_value, **kwargs)
 
-        types_signature: str = get_types_signature(updated_types, updated_decorator, execution_space)
+        is_standalone_workunit: bool = True
+        list_passed: bool = True
+
+        if not isinstance(workunit, list): 
+            workunit = [workunit] # for easier transformations
+            list_passed = False
+
+        parser = self.compiler.get_parser(get_metadata(workunit[0]).path)
+        entity_AST = []
+
+        for this_workunit in workunit:
+            this_metadata = get_metadata(this_workunit)
+            this_tree = parser.get_entity(this_metadata.name).AST
+            workunit_str = str(this_workunit)
+
+            if not isinstance(this_tree, ast.FunctionDef):
+                # Workunit must be a function on its own (not in a functor)
+                is_standalone_workunit = False
+                entity_AST.append(this_tree)
+                continue
+
+            is_missing_annotations: bool = (
+                workunit_str in self.workunit_params
+                or
+                list_passed
+                or
+                check_missing_annotations(this_tree.args.args)
+                )
+
+            if is_missing_annotations:
+                if workunit_str in self.workunit_params:
+                    this_tree.args.args = pickle.loads(self.workunit_params[workunit_str])
+                else:
+                    # first call, store the original params
+                    self.workunit_params[workunit_str] = pickle.dumps(this_tree.args.args)
+            
+            entity_AST.append(this_tree)
+
+        workunit_tree_tup: Union[List[Tuple[Callable, ast.AST]], Tuple[Callable, ast.AST]]
+
+        if not list_passed: # revert to singular tuple if not list originally
+            workunit_tree_tup = (workunit[0], entity_AST[0])
+            workunit = workunit[0]
+            entity_AST = None # No fusion
+        else:
+            workunit_tree_tup = list(zip(workunit, entity_AST))
+
+        updated_types: UpdatedTypes = None
+        updated_decorator: UpdatedDecorator = None
+        types_signature: str = None
+
+        if is_standalone_workunit:
+            if is_missing_annotations:
+                updated_types = get_annotations(f"parallel_{operation}", workunit_tree_tup, policy, passed_kwargs=kwargs)
+            updated_decorator = get_views_decorator(workunit_tree_tup, passed_kwargs=kwargs)
+            types_signature = get_types_signature(updated_types, updated_decorator, execution_space)
+
```

Comment:
You should move everything you added to `run_workunit` to `args_type_inference.py`. This is too much detail for the runtime, it should be replaces by a single method call

Reply:
I tried doing that initially, but I realized we will need some design changes. The biggest one is where do we keep the cache for the original arg nodes. It made sense to me to keep them at `Runtime`, since the compiler and parser are accessibly by design here.

Now I can move this to `args_type_inference.py` but that will require, either

1.  Moving the cache there 
2. Interfacing with runtime object from `args_type_inference.py`, which in my opinion does not feel right (?) and will look convoluted.

I have however, abstracted out some parts of the details. Let me know if you think this is ok. Otherwise, I am happy to move the cache to `args_type_inference.py` as a global (or let me know if you see another elegant solution)

Reply:
If you're referring to `self.workunit_params` as the cache, I think it makes sense to move it to args_type_inference.py